fncas/fncas_optimize.h

```diff
@@ -0,0 +1,266 @@
+    grad_eps_ = params.GetValue("grad_eps", grad_eps_);
+  }
+  OptimizationResult Optimize(const std::vector<double>& starting_point);
+private:
+  size_t max_steps_ = 5000;
+  double bt_alpha_ = 0.5;
+  double bt_beta_ = 0.8;
+  size_t bt_max_steps_ = 100;
+  double grad_eps_ = 1e-8;
+};
+
+template <class F>
+class ConjugateGradientOptimizer : noncopyable {
+public:
+  ConjugateGradientOptimizer() { }
+  ConjugateGradientOptimizer(OptimizerParameters &params) {
+    max_steps_ = params.GetValue("max_steps", max_steps_);
+    bt_alpha_ = params.GetValue("bt_alpha", bt_alpha_);
+    bt_beta_ = params.GetValue("bt_beta", bt_beta_);
+    bt_max_steps_ = params.GetValue("bt_max_steps", bt_max_steps_);
+    grad_eps_ = params.GetValue("grad_eps", grad_eps_);
+  }
+  OptimizationResult Optimize(const std::vector<double>& starting_point);
+private:
+  size_t max_steps_ = 5000;
+  double bt_alpha_ = 0.5;
+  double bt_beta_ = 0.8;
+  size_t bt_max_steps_ = 100;
+  double grad_eps_ = 1e-8;
+
+  double PolakRibiere(const std::vector<double> &g, const std::vector<double> &g_prev);
+};
+
+template<class F> 
+OptimizationResult GradientDescentOptimizer<F>::Optimize(const std::vector<double>& starting_point) {
+  fncas::reset_internals_singleton();
+  const size_t dim = starting_point.size();
+  const fncas::x gradient_helper(dim);
+  fncas::f_intermediate fi(F::compute(gradient_helper));
+  fncas::g_intermediate gi = fncas::g_intermediate(gradient_helper, fi);
+  std::vector<double> current_point(starting_point);
+
+  for (size_t iteration = 0; iteration < max_steps_; ++iteration) {
+    const auto g = gi(current_point);
+    const auto try_step = [&dim, &current_point, &fi, &g](double step) -> std::pair<double, std::vector<double>> {
+      std::vector<double> candidate(dim);
+      for (size_t i = 0; i < dim; ++i) {
+        candidate[i] = current_point[i] - g.gradient[i] * step;
+      }
+      const double value = fi(candidate);
+      return std::make_pair(value, candidate);
+    };
+    current_point = std::min(try_step(0.01), std::min(try_step(0.05), try_step(0.2))).second;
+  }
+
+  return OptimizationResult{fi(current_point), current_point};
+}
+
+
+template<class F> 
+OptimizationResult GradientDescentOptimizerBT<F>::Optimize(const std::vector<double>& starting_point) {
+  fncas::reset_internals_singleton();
+  const size_t dim = starting_point.size();
+  const fncas::x gradient_helper(dim);
+  fncas::f_intermediate fi(F::compute(gradient_helper));
+  fncas::g_intermediate gi = fncas::g_intermediate(gradient_helper, fi);
+  std::vector<double> current_point(starting_point);
+
+  for (size_t iteration = 0; iteration < max_steps_; ++iteration) {
+    const auto current_f_gradf = gi(current_point);
+    const double current_f_value = current_f_gradf.value;
+    double squared_gradient = 0.0;
+    std::vector<double> test_point(dim);
+
+    for (size_t i = 0; i < dim; ++i) {
+      test_point[i] = current_point[i] - current_f_gradf.gradient[i];
```

Comment:
FTR, I'm **really** scared of this. With many exponential functions, making a step of a full gradient length takes the value to `infinity`, which is non-recoverable, unless extra logic handles that case specifically.